{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "39f040e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Jill E. Moore\n",
    "#Moore Lab - UMass Chan Medical School\n",
    "#May 2022\n",
    "\n",
    "import os\n",
    "import json\n",
    "import urllib\n",
    "from urllib import request, parse, error\n",
    "import requests\n",
    "import base64\n",
    "from statistics import mean, median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b88b08aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Process_Token():\n",
    "    credentials=open(\"/home/moorej3/.encode.txt\")\n",
    "    credArray=next(credentials).rstrip().split(\"\\t\")\n",
    "    return credArray[0], credArray[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "adf7a13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Extract_Experiment_Metatdata(dataset, creds, genome): #extracts experiment metadata\n",
    "    try:\n",
    "        dataDir=\"/data/projects/encode/json/exps/\"+dataset\n",
    "        json_data=open(dataDir+\".json\").read()\n",
    "        data = json.loads(json_data)\n",
    "    except:\n",
    "        url = \"https://www.encodeproject.org/\"+dataset+\"/?format=json\"\n",
    "        request = urllib.request.Request(url)\n",
    "        request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "        response = urllib.request.urlopen(request)\n",
    "        data = json.loads(response.read())\n",
    "\n",
    "    assayTermName = data[\"assay_term_name\"]\n",
    "    assayType = data[\"assay_slims\"][0]\n",
    "    \n",
    "    try:\n",
    "        biosample = data[\"biosample_summary\"].replace(\"nuclear fraction \",\"\")\n",
    "    except:\n",
    "        biosample = \"NA\"\n",
    "        \n",
    "    try:\n",
    "        tissue = data[\"biosample_ontology\"][\"term_name\"]\n",
    "    except:\n",
    "        \n",
    "        url = \"https://www.encodeproject.org/\"+dataset+\"/?format=json\"\n",
    "        request = urllib.request.Request(url)\n",
    "        request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "        response = urllib.request.urlopen(request)\n",
    "        data = json.loads(response.read())\n",
    "        \n",
    "        tissue = data[\"biosample_ontology\"][\"term_name\"]\n",
    "    \n",
    "    donors = []\n",
    "    treatments = []\n",
    "    for entry in data[\"replicates\"]:\n",
    "        donors.append(entry[\"library\"][\"biosample\"][\"donor\"][\"accession\"])\n",
    "        try:\n",
    "            for t in entry[\"library\"][\"biosample\"][\"treatments\"]:\n",
    "                treatments.append(t[\"treatment_term_name\"])\n",
    "        except:\n",
    "            pass \n",
    "    donor = \",\".join(list(set(donors)))\n",
    "    treatment = \",\".join(list(set(treatments)))\n",
    "    \n",
    "    status=data[\"status\"]\n",
    "    \n",
    "    targets = []\n",
    "    if \"target\" in data:\n",
    "        targets.append(data[\"target\"][\"label\"])\n",
    "    \n",
    "    numberReads = []\n",
    "    \n",
    "    for entry in data[\"files\"]: #loops through files associated with experiment\n",
    "        try:\n",
    "            if entry[\"file_type\"] == \"fastq\" and entry[\"status\"] == \"released\":\n",
    "\n",
    "                replicate = \",\".join([str(i) for i in entry[\"biological_replicates\"]])\n",
    "                numReads = entry[\"read_count\"]\n",
    "                numberReads.append(numReads)\n",
    "        except:\n",
    "            print(dataset)\n",
    "    numberReads = list(set(numberReads))\n",
    "    return assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "979e6d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "usrname, psswd = Process_Token()\n",
    "base64string = base64.b64encode(bytes('%s:%s' % (usrname,psswd),'ascii'))\n",
    "creds = base64string.decode('utf-8')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "2eb8d983",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCSR917JIA\n",
      "ENCSR917JIA\n",
      "ENCSR724HQB\n",
      "ENCSR724HQB\n",
      "Total # biosamples \t 915\n",
      "Total # cell/tissue types\t 284\n",
      "DNA binding\n",
      "\t ChIP-seq \t 1537 \t 211 \t 73 \t 16815894 \t 126875918514 \t 613\n",
      "\t Mint-ChIP-seq \t 872 \t 156 \t 63 \t 3259418 \t 58646709053 \t 6\n",
      "3D chromatin structure\n",
      "\t ChIA-PET \t 70 \t 37 \t 37 \t 315755020 \t 41363907612 \t 2\n",
      "\t HiC \t 33 \t 21 \t 14 \t 134140592 \t 39303193455 \t 0\n",
      "RNA binding\n",
      "\t eCLIP \t 55 \t 2 \t 2 \t 19095308 \t 2272341709 \t 44\n",
      "DNA accessibility\n",
      "\t ATAC-seq \t 251 \t 237 \t 171 \t 41849455 \t 36785670826 \t 0\n",
      "\t DNase-seq \t 661 \t 575 \t 137 \t 48941770 \t 65679855283 \t 0\n",
      "Single cell\n",
      "\t single-cell RNA sequencing assay \t 75 \t 75 \t 31 \t 127617558 \t 19142633648 \t 0\n",
      "\t single-nucleus ATAC-seq \t 181 \t 155 \t 87 \t 134540532 \t 105345236771 \t 0\n",
      "Transcription\n",
      "\t RNA-seq \t 219 \t 179 \t 82 \t 35719000 \t 14109005085 \t 0\n",
      "\t microRNA-seq \t 101 \t 96 \t 72 \t 16837154 \t 2576084568 \t 0\n",
      "\t long read RNA-seq \t 98 \t 86 \t 55 \t 1550514 \t 218622429 \t 0\n",
      "\t Bru-seq \t 17 \t 17 \t 17 \t 62350989 \t 2119933636 \t 0\n",
      "\t BruUV-seq \t 16 \t 16 \t 16 \t 62642963 \t 2004574804 \t 0\n",
      "\t BruChase-seq \t 32 \t 16 \t 16 \t 60117407 \t 3847514022 \t 0\n",
      "\t polyA plus RNA-seq \t 54 \t 54 \t 9 \t 30594025 \t 1652077364 \t 0\n",
      "\t PRO-cap \t 6 \t 1 \t 1 \t 15199946 \t 273599021 \t 0\n",
      "\t PAS-seq \t 4 \t 4 \t 4 \t 95374180 \t 1049115980 \t 0\n",
      "\t PRO-seq \t 2 \t 1 \t 1 \t 30873075 \t 123492301 \t 0\n",
      "\t GRO-cap \t 4 \t 2 \t 2 \t 40028887 \t 160115548 \t 0\n",
      "DNA methylation\n",
      "\t whole-genome shotgun bisulfite sequencing \t 10 \t 6 \t 1 \t 146970751 \t 5878830056 \t 0\n"
     ]
    }
   ],
   "source": [
    "## Parameters\n",
    "genome = \"GRCh38\"\n",
    "species = \"Homo+sapiens\"\n",
    "\n",
    "## Build query\n",
    "urlMain = \"https://www.encodeproject.org/search/?type=Experiment&\" + \\\n",
    "    \"control_type!=*&status=released&perturbed=false&award.rfa=ENCODE4&\" +\\\n",
    "    \"replicates.library.biosample.donor.organism.scientific_name=\" + species + \\\n",
    "    \"&format=json&limit=all\" #update limit=all for all datasets\n",
    "\n",
    "request = urllib.request.Request(urlMain)\n",
    "request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "response = urllib.request.urlopen(request)\n",
    "data = json.loads(response.read())\n",
    "\n",
    "biosampleList = []\n",
    "tissueList = []\n",
    "experimentDict = {}\n",
    "\n",
    "for entry in data[\"@graph\"]: #loops through experiments\n",
    "    experiment = entry[\"accession\"]\n",
    "    assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets = \\\n",
    "        Extract_Experiment_Metatdata(experiment, creds, genome)\n",
    "    if assayType not in experimentDict:\n",
    "        experimentDict[assayType] = {}\n",
    "    if assayTermName not in experimentDict[assayType]:\n",
    "        experimentDict[assayType][assayTermName] = {\"numExp\": 0, \"numRead\":[], \"biosamples\":[], \"tissues\":[], \"biosampleType\":{}, \"targets\":[]}\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"numExp\"] += 1\n",
    "    experimentDict[assayType][assayTermName][\"numRead\"] += numberReads\n",
    "    experimentDict[assayType][assayTermName][\"targets\"] += targets\n",
    "    \n",
    "    longBiosample = tissue + \" \" + treatment + \" \" + donor\n",
    "    experimentDict[assayType][assayTermName][\"biosamples\"].append(longBiosample)\n",
    "    biosampleList.append(longBiosample)\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"tissues\"].append(tissue)\n",
    "    tissueList.append(tissue)\n",
    "\n",
    "print(\"Total # biosamples\", \"\\t\", len(list(set(biosampleList))))\n",
    "print(\"Total # cell/tissue types\" \"\\t\", len(list(set(tissueList))))\n",
    "\n",
    "for x in experimentDict:\n",
    "    print(x)\n",
    "    for y in experimentDict[x]:\n",
    "        data = experimentDict[x][y]\n",
    "        meanRead = round(mean(data[\"numRead\"]))\n",
    "        sumRead = sum(data[\"numRead\"])\n",
    "        numTargets = len(list(set(data[\"targets\"])))\n",
    "        print(\"\\t\", y, \"\\t\", data[\"numExp\"], \"\\t\", len(list(set(data[\"biosamples\"]))), \"\\t\", len(list(set(data[\"tissues\"]))), \"\\t\", meanRead, \"\\t\", sumRead, \"\\t\", numTargets)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fcbead4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
