{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "60e47460",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Jill E. Moore\n",
    "#Moore Lab - UMass Chan Medical School\n",
    "#May 2022\n",
    "\n",
    "import os\n",
    "import json\n",
    "import urllib\n",
    "from urllib import request, parse, error\n",
    "import requests\n",
    "import base64\n",
    "from statistics import mean, median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "413e07d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Process_Token():\n",
    "    credentials=open(\"/home/moorej3/.encode.txt\")\n",
    "    credArray=next(credentials).rstrip().split(\"\\t\")\n",
    "    return credArray[0], credArray[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef601ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Extract_Experiment_Metatdata(dataset, creds): #extracts experiment metadata\n",
    "    try:\n",
    "        dataDir=\"/data/projects/encode/json/exps/\"+dataset\n",
    "        json_data=open(dataDir+\".json\").read()\n",
    "        data = json.loads(json_data)\n",
    "    except:\n",
    "        url = \"https://www.encodeproject.org/\"+dataset+\"/?format=json\"\n",
    "        request = urllib.request.Request(url)\n",
    "        request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "        response = urllib.request.urlopen(request)\n",
    "        data = json.loads(response.read())\n",
    "\n",
    "    #assayTermName = data[\"assay_term_name\"]\n",
    "    assayTermName = data[\"assay_title\"]\n",
    "    assayType = data[\"assay_slims\"][0]\n",
    "    \n",
    "    try:\n",
    "        biosample = data[\"biosample_summary\"].replace(\"nuclear fraction \",\"\")\n",
    "    except:\n",
    "        biosample = \"NA\"\n",
    "        \n",
    "    try:\n",
    "        tissue = data[\"biosample_ontology\"][\"term_name\"]\n",
    "    except:\n",
    "        \n",
    "        url = \"https://www.encodeproject.org/\"+dataset+\"/?format=json\"\n",
    "        request = urllib.request.Request(url)\n",
    "        request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "        response = urllib.request.urlopen(request)\n",
    "        data = json.loads(response.read())\n",
    "        \n",
    "        tissue = data[\"biosample_ontology\"][\"term_name\"]\n",
    "    \n",
    "    donors = []\n",
    "    treatments = []\n",
    "    for entry in data[\"replicates\"]:\n",
    "        donors.append(entry[\"library\"][\"biosample\"][\"donor\"][\"accession\"])\n",
    "        try:\n",
    "            for t in entry[\"library\"][\"biosample\"][\"treatments\"]:\n",
    "                treatments.append(t[\"treatment_term_name\"])\n",
    "        except:\n",
    "            pass \n",
    "    donor = \",\".join(list(set(donors)))\n",
    "    treatment = \",\".join(list(set(treatments)))\n",
    "    \n",
    "    status=data[\"status\"]\n",
    "    \n",
    "    targets = []\n",
    "    if \"target\" in data:\n",
    "        targets.append(data[\"target\"][\"label\"])\n",
    "    \n",
    "    numberReads = []\n",
    "    \n",
    "    for entry in data[\"files\"]: #loops through files associated with experiment\n",
    "        try:\n",
    "            if entry[\"file_type\"] == \"fastq\" and entry[\"status\"] == \"released\":\n",
    "\n",
    "                replicate = \",\".join([str(i) for i in entry[\"biological_replicates\"]])\n",
    "                numReads = entry[\"read_count\"]\n",
    "                numberReads.append(numReads)\n",
    "        except:\n",
    "            print(dataset)\n",
    "    #numberReads = list(set(numberReads))\n",
    "    return assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8ceb0844",
   "metadata": {},
   "outputs": [],
   "source": [
    "usrname, psswd = Process_Token()\n",
    "base64string = base64.b64encode(bytes('%s:%s' % (usrname,psswd),'ascii'))\n",
    "creds = base64string.decode('utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0b0cf00",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "93603f07",
   "metadata": {},
   "source": [
    "## ENCODE Phase IV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03324445",
   "metadata": {},
   "source": [
    "### ENCODE4 - Human & Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1cb6edee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCSR724HQB\n",
      "ENCSR724HQB\n",
      "ENCSR917JIA\n",
      "ENCSR917JIA\n",
      "Total # biosamples \t 1341\n",
      "Total # cell/tissue types\t 318\n",
      "DNA binding\n",
      "\t Mint-ChIP-seq \t 977 \t 154 \t 62 \t 131857097140 \t 6\n",
      "\t Histone ChIP-seq \t 791 \t 194 \t 53 \t 74893183558 \t 6\n",
      "\t TF ChIP-seq \t 875 \t 191 \t 74 \t 92316660531 \t 619\n",
      "CRISPR screen\n",
      "\t Flow-FISH CRISPR screen \t 280 \t 9 \t 5 \t 4847280143 \t 0\n",
      "\t proliferation CRISPR screen \t 53 \t 13 \t 11 \t 4351285171 \t 0\n",
      "\t FACS CRISPR screen \t 12 \t 7 \t 2 \t 965222911 \t 0\n",
      "Transcription\n",
      "\t CRISPR RNA-seq \t 336 \t 2 \t 2 \t 81858461506 \t 215\n",
      "\t total RNA-seq \t 452 \t 367 \t 96 \t 39449715780 \t 0\n",
      "\t microRNA-seq \t 197 \t 191 \t 80 \t 5049178324 \t 0\n",
      "\t shRNA RNA-seq \t 23 \t 2 \t 2 \t 2238265452 \t 18\n",
      "\t polyA plus RNA-seq \t 56 \t 56 \t 11 \t 3830639560 \t 0\n",
      "\t Bru-seq \t 49 \t 34 \t 17 \t 13159190800 \t 0\n",
      "\t BruChase-seq \t 32 \t 16 \t 16 \t 7695028044 \t 0\n",
      "\t BruUV-seq \t 16 \t 16 \t 16 \t 4009149608 \t 0\n",
      "\t PAS-seq \t 7 \t 7 \t 7 \t 1424618846 \t 0\n",
      "\t PRO-seq \t 20 \t 4 \t 3 \t 3167532333 \t 0\n",
      "\t PRO-cap \t 25 \t 8 \t 7 \t 4386600070 \t 0\n",
      "\t long read RNA-seq \t 162 \t 146 \t 63 \t 385917982 \t 0\n",
      "\t GRO-cap \t 4 \t 2 \t 2 \t 160115548 \t 0\n",
      "Massively parallel reporter assay\n",
      "\t MPRA \t 27 \t 7 \t 6 \t 16441215941 \t 0\n",
      "\t STARR-seq \t 21 \t 17 \t 6 \t 26997995089 \t 0\n",
      "DNA accessibility\n",
      "\t DNase-seq \t 890 \t 779 \t 147 \t 223332152226 \t 0\n",
      "\t ATAC-seq \t 368 \t 302 \t 183 \t 98254548592 \t 0\n",
      "3D chromatin structure\n",
      "\t ChIA-PET \t 136 \t 58 \t 50 \t 182990818934 \t 2\n",
      "\t Hi-C \t 59 \t 40 \t 20 \t 151374854320 \t 0\n",
      "Single cell\n",
      "\t long read scRNA-seq \t 3 \t 3 \t 2 \t 2277781 \t 0\n",
      "\t perturbation followed by scRNA-seq \t 1 \t 1 \t 1 \t 22825038768 \t 0\n",
      "\t scRNA-seq \t 348 \t 180 \t 42 \t 77160421370 \t 0\n",
      "\t snATAC-seq \t 224 \t 168 \t 94 \t 301935382514 \t 0\n",
      "RNA binding\n",
      "\t eCLIP \t 55 \t 2 \t 2 \t 4544683418 \t 44\n",
      "RNA structure\n",
      "\t icSHAPE \t 3 \t 3 \t 3 \t 1900847958 \t 0\n",
      "\t icLASER \t 1 \t 1 \t 1 \t 1368276853 \t 0\n",
      "DNA methylation\n",
      "\t WGBS \t 10 \t 6 \t 1 \t 11757660112 \t 0\n"
     ]
    }
   ],
   "source": [
    "## Parameters\n",
    "species1 = \"Homo+sapiens\"\n",
    "species2 = \"Mus+musculus\"\n",
    "rfa=\"ENCODE4\"\n",
    "\n",
    "## Build query\n",
    "urlMain = \"https://www.encodeproject.org/search/?type=Experiment&\" + \\\n",
    "    \"type=FunctionalCharacterizationExperiment&\" + \\\n",
    "    \"control_type!=*&status=released&award.rfa=\" + rfa + \\\n",
    "    \"&replicates.library.biosample.donor.organism.scientific_name=\" + species1 + \\\n",
    "    \"&replicates.library.biosample.donor.organism.scientific_name=\" + species2 + \\\n",
    "    \"&format=json&limit=all\" #update limit=all for all datasets\n",
    "\n",
    "request = urllib.request.Request(urlMain)\n",
    "request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "response = urllib.request.urlopen(request)\n",
    "data = json.loads(response.read())\n",
    "\n",
    "biosampleList = []\n",
    "tissueList = []\n",
    "experimentDict = {}\n",
    "\n",
    "for entry in data[\"@graph\"]: #loops through experiments\n",
    "    experiment = entry[\"accession\"]\n",
    "    assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets = \\\n",
    "        Extract_Experiment_Metatdata(experiment, creds)\n",
    "    if assayType not in experimentDict:\n",
    "        experimentDict[assayType] = {}\n",
    "    if assayTermName not in experimentDict[assayType]:\n",
    "        experimentDict[assayType][assayTermName] = {\"numExp\": 0, \"numRead\":[], \"biosamples\":[], \"tissues\":[], \"biosampleType\":{}, \"targets\":[]}\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"numExp\"] += 1\n",
    "    experimentDict[assayType][assayTermName][\"numRead\"] += numberReads\n",
    "    experimentDict[assayType][assayTermName][\"targets\"] += targets\n",
    "    \n",
    "    if \"Homo sapiens\" in biosample:\n",
    "        longBiosample = tissue + \" \" + treatment + \" \" + donor\n",
    "    elif \"Mus musculus\" in biosample:\n",
    "        longBiosample = biosample\n",
    "    else:\n",
    "        longBiosample = biosample\n",
    "    experimentDict[assayType][assayTermName][\"biosamples\"].append(longBiosample)\n",
    "    biosampleList.append(longBiosample)\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"tissues\"].append(tissue)\n",
    "    tissueList.append(tissue)\n",
    "\n",
    "print(\"Total # biosamples\", \"\\t\", len(list(set(biosampleList))))\n",
    "print(\"Total # cell/tissue types\" \"\\t\", len(list(set(tissueList))))\n",
    "\n",
    "for x in experimentDict:\n",
    "    print(x)\n",
    "    for y in experimentDict[x]:\n",
    "        data = experimentDict[x][y]\n",
    "        sumRead = sum(data[\"numRead\"])\n",
    "        numTargets = len(list(set(data[\"targets\"])))\n",
    "        print(\"\\t\", y, \"\\t\", data[\"numExp\"], \"\\t\", len(list(set(data[\"biosamples\"]))), \"\\t\", len(list(set(data[\"tissues\"]))), \"\\t\", sumRead, \"\\t\", numTargets)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7ef80fb",
   "metadata": {},
   "source": [
    "### ENCODE4 - Human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8a303e9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCSR724HQB\n",
      "ENCSR724HQB\n",
      "ENCSR917JIA\n",
      "ENCSR917JIA\n",
      "Total # biosamples \t 1067\n",
      "Total # cell/tissue types\t 293\n",
      "CRISPR screen\n",
      "\t Flow-FISH CRISPR screen \t 280 \t 9 \t 5 \t 4847280143 \t 0\n",
      "\t proliferation CRISPR screen \t 53 \t 13 \t 11 \t 4351285171 \t 0\n",
      "\t FACS CRISPR screen \t 2 \t 1 \t 1 \t 44655680 \t 0\n",
      "DNA accessibility\n",
      "\t ATAC-seq \t 368 \t 302 \t 183 \t 98254548592 \t 0\n",
      "\t DNase-seq \t 713 \t 603 \t 136 \t 146623587892 \t 0\n",
      "DNA binding\n",
      "\t Mint-ChIP-seq \t 977 \t 154 \t 62 \t 131857097140 \t 6\n",
      "\t Histone ChIP-seq \t 791 \t 194 \t 53 \t 74893183558 \t 6\n",
      "\t TF ChIP-seq \t 875 \t 191 \t 74 \t 92316660531 \t 619\n",
      "Massively parallel reporter assay\n",
      "\t MPRA \t 27 \t 7 \t 6 \t 16441215941 \t 0\n",
      "\t STARR-seq \t 21 \t 17 \t 6 \t 26997995089 \t 0\n",
      "Transcription\n",
      "\t total RNA-seq \t 373 \t 288 \t 91 \t 30815536968 \t 0\n",
      "\t CRISPR RNA-seq \t 336 \t 2 \t 2 \t 81858461506 \t 215\n",
      "\t microRNA-seq \t 108 \t 103 \t 73 \t 2662795485 \t 0\n",
      "\t shRNA RNA-seq \t 23 \t 2 \t 2 \t 2238265452 \t 18\n",
      "\t Bru-seq \t 49 \t 34 \t 17 \t 13159190800 \t 0\n",
      "\t BruChase-seq \t 32 \t 16 \t 16 \t 7695028044 \t 0\n",
      "\t BruUV-seq \t 16 \t 16 \t 16 \t 4009149608 \t 0\n",
      "\t PAS-seq \t 4 \t 4 \t 4 \t 1049115980 \t 0\n",
      "\t polyA plus RNA-seq \t 54 \t 54 \t 9 \t 3304154728 \t 0\n",
      "\t PRO-seq \t 20 \t 4 \t 3 \t 3167532333 \t 0\n",
      "\t GRO-cap \t 4 \t 2 \t 2 \t 160115548 \t 0\n",
      "\t PRO-cap \t 25 \t 8 \t 7 \t 4386600070 \t 0\n",
      "\t long read RNA-seq \t 98 \t 86 \t 55 \t 218622429 \t 0\n",
      "Single cell\n",
      "\t scRNA-seq \t 75 \t 75 \t 31 \t 38285267296 \t 0\n",
      "\t perturbation followed by scRNA-seq \t 1 \t 1 \t 1 \t 22825038768 \t 0\n",
      "\t snATAC-seq \t 181 \t 155 \t 87 \t 264219800861 \t 0\n",
      "3D chromatin structure\n",
      "\t ChIA-PET \t 125 \t 52 \t 47 \t 168187967516 \t 2\n",
      "\t Hi-C \t 43 \t 25 \t 15 \t 106673987860 \t 0\n",
      "RNA binding\n",
      "\t eCLIP \t 55 \t 2 \t 2 \t 4544683418 \t 44\n",
      "RNA structure\n",
      "\t icSHAPE \t 3 \t 3 \t 3 \t 1900847958 \t 0\n",
      "\t icLASER \t 1 \t 1 \t 1 \t 1368276853 \t 0\n",
      "DNA methylation\n",
      "\t WGBS \t 10 \t 6 \t 1 \t 11757660112 \t 0\n"
     ]
    }
   ],
   "source": [
    "## Parameters\n",
    "species = \"Homo+sapiens\"\n",
    "rfa=\"ENCODE4\"\n",
    "\n",
    "## Build query\n",
    "urlMain = \"https://www.encodeproject.org/search/?type=Experiment&\" + \\\n",
    "    \"type=FunctionalCharacterizationExperiment&\" + \\\n",
    "    \"control_type!=*&status=released&award.rfa=\" + rfa + \\\n",
    "    \"&replicates.library.biosample.donor.organism.scientific_name=\" + species + \\\n",
    "    \"&format=json&limit=all\" #update limit=all for all datasets\n",
    "\n",
    "request = urllib.request.Request(urlMain)\n",
    "request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "response = urllib.request.urlopen(request)\n",
    "data = json.loads(response.read())\n",
    "\n",
    "biosampleList = []\n",
    "tissueList = []\n",
    "experimentDict = {}\n",
    "\n",
    "for entry in data[\"@graph\"]: #loops through experiments\n",
    "    experiment = entry[\"accession\"]\n",
    "    assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets = \\\n",
    "        Extract_Experiment_Metatdata(experiment, creds)\n",
    "    if assayType not in experimentDict:\n",
    "        experimentDict[assayType] = {}\n",
    "    if assayTermName not in experimentDict[assayType]:\n",
    "        experimentDict[assayType][assayTermName] = {\"numExp\": 0, \"numRead\":[], \"biosamples\":[], \"tissues\":[], \"biosampleType\":{}, \"targets\":[]}\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"numExp\"] += 1\n",
    "    experimentDict[assayType][assayTermName][\"numRead\"] += numberReads\n",
    "    experimentDict[assayType][assayTermName][\"targets\"] += targets\n",
    "    \n",
    "    if \"Homo sapiens\" in biosample:\n",
    "        longBiosample = tissue + \" \" + treatment + \" \" + donor\n",
    "    elif \"Mus musculus\" in biosample:\n",
    "        longBiosample = biosample\n",
    "    else:\n",
    "        longBiosample = biosample\n",
    "    experimentDict[assayType][assayTermName][\"biosamples\"].append(longBiosample)\n",
    "    biosampleList.append(longBiosample)\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"tissues\"].append(tissue)\n",
    "    tissueList.append(tissue)\n",
    "\n",
    "print(\"Total # biosamples\", \"\\t\", len(list(set(biosampleList))))\n",
    "print(\"Total # cell/tissue types\" \"\\t\", len(list(set(tissueList))))\n",
    "\n",
    "for x in experimentDict:\n",
    "    print(x)\n",
    "    for y in experimentDict[x]:\n",
    "        data = experimentDict[x][y]\n",
    "        sumRead = sum(data[\"numRead\"])\n",
    "        numTargets = len(list(set(data[\"targets\"])))\n",
    "        print(\"\\t\", y, \"\\t\", data[\"numExp\"], \"\\t\", len(list(set(data[\"biosamples\"]))), \"\\t\", len(list(set(data[\"tissues\"]))), \"\\t\", sumRead, \"\\t\", numTargets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63789a47",
   "metadata": {},
   "source": [
    "### ENCODE4 - Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b3604a98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total # biosamples \t 274\n",
      "Total # cell/tissue types\t 32\n",
      "CRISPR screen\n",
      "\t FACS CRISPR screen \t 10 \t 6 \t 1 \t 920567231 \t 0\n",
      "Transcription\n",
      "\t microRNA-seq \t 89 \t 88 \t 8 \t 2386382839 \t 0\n",
      "\t total RNA-seq \t 79 \t 79 \t 6 \t 8634178812 \t 0\n",
      "\t PAS-seq \t 3 \t 3 \t 3 \t 375502866 \t 0\n",
      "\t long read RNA-seq \t 64 \t 60 \t 9 \t 167295553 \t 0\n",
      "\t polyA plus RNA-seq \t 2 \t 2 \t 2 \t 526484832 \t 0\n",
      "DNA accessibility\n",
      "\t DNase-seq \t 177 \t 176 \t 17 \t 76708564334 \t 0\n",
      "Single cell\n",
      "\t snATAC-seq \t 43 \t 13 \t 8 \t 37715581653 \t 0\n",
      "\t scRNA-seq \t 273 \t 105 \t 12 \t 38875154074 \t 0\n",
      "\t long read scRNA-seq \t 3 \t 3 \t 2 \t 2277781 \t 0\n",
      "3D chromatin structure\n",
      "\t ChIA-PET \t 11 \t 6 \t 3 \t 14802851418 \t 2\n",
      "\t Hi-C \t 16 \t 15 \t 5 \t 44700866460 \t 0\n"
     ]
    }
   ],
   "source": [
    "## Parameters\n",
    "species = \"Mus+musculus\"\n",
    "rfa=\"ENCODE4\"\n",
    "\n",
    "## Build query\n",
    "urlMain = \"https://www.encodeproject.org/search/?type=Experiment&\" + \\\n",
    "    \"type=FunctionalCharacterizationExperiment&\" + \\\n",
    "    \"control_type!=*&status=released&award.rfa=\" + rfa + \\\n",
    "    \"&replicates.library.biosample.donor.organism.scientific_name=\" + species + \\\n",
    "    \"&format=json&limit=all\" #update limit=all for all datasets\n",
    "\n",
    "request = urllib.request.Request(urlMain)\n",
    "request.add_header(\"Authorization\", \"Basic %s\" % creds)\n",
    "response = urllib.request.urlopen(request)\n",
    "data = json.loads(response.read())\n",
    "\n",
    "biosampleList = []\n",
    "tissueList = []\n",
    "experimentDict = {}\n",
    "\n",
    "for entry in data[\"@graph\"]: #loops through experiments\n",
    "    experiment = entry[\"accession\"]\n",
    "    assayTermName, assayType, biosample, tissue, donor, treatment, status, numberReads, targets = \\\n",
    "        Extract_Experiment_Metatdata(experiment, creds)\n",
    "    if assayType not in experimentDict:\n",
    "        experimentDict[assayType] = {}\n",
    "    if assayTermName not in experimentDict[assayType]:\n",
    "        experimentDict[assayType][assayTermName] = {\"numExp\": 0, \"numRead\":[], \"biosamples\":[], \"tissues\":[], \"biosampleType\":{}, \"targets\":[]}\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"numExp\"] += 1\n",
    "    experimentDict[assayType][assayTermName][\"numRead\"] += numberReads\n",
    "    experimentDict[assayType][assayTermName][\"targets\"] += targets\n",
    "    \n",
    "    if \"Homo sapiens\" in biosample:\n",
    "        longBiosample = tissue + \" \" + treatment + \" \" + donor\n",
    "    elif \"Mus musculus\" in biosample:\n",
    "        longBiosample = biosample\n",
    "    else:\n",
    "        longBiosample = biosample\n",
    "    experimentDict[assayType][assayTermName][\"biosamples\"].append(longBiosample)\n",
    "    biosampleList.append(longBiosample)\n",
    "    \n",
    "    experimentDict[assayType][assayTermName][\"tissues\"].append(tissue)\n",
    "    tissueList.append(tissue)\n",
    "\n",
    "print(\"Total # biosamples\", \"\\t\", len(list(set(biosampleList))))\n",
    "print(\"Total # cell/tissue types\" \"\\t\", len(list(set(tissueList))))\n",
    "\n",
    "for x in experimentDict:\n",
    "    print(x)\n",
    "    for y in experimentDict[x]:\n",
    "        data = experimentDict[x][y]\n",
    "        sumRead = sum(data[\"numRead\"])\n",
    "        numTargets = len(list(set(data[\"targets\"])))\n",
    "        print(\"\\t\", y, \"\\t\", data[\"numExp\"], \"\\t\", len(list(set(data[\"biosamples\"]))), \"\\t\", len(list(set(data[\"tissues\"]))), \"\\t\", sumRead, \"\\t\", numTargets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f954eb0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#urlMain = \"https://www.encodeproject.org/search/?type=Experiment&\" + \\\n",
    "#    \"type=FunctionalCharacterizationExperiment&\" + \\\n",
    "#    \"control_type!=*&status=released&perturbed=true&perturbed=false\" + \\\n",
    "#    \"&award.rfa=ENCODE4&award.rfa=ENCODE3&award.rfa=ENCODE2&award.rfa=ENCODE2-Mouse\" + \\\n",
    "#    \"&replicates.library.biosample.donor.organism.scientific_name=\" + species1 + \\\n",
    "#    \"&format=json&limit=all\" #update limit=all for all datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9eecc6e",
   "metadata": {},
   "source": [
    "## Previous ENCODE Versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133c4769",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
